File: ingestor/textos.py
"""Normalização e busca de trechos em texto livre (regulamentos).

A normalização preserva comprimento: cada caractere do original vira exatamente
um caractere no texto normalizado. Isso deixa os offsets alinhados entre `texto`
e `texto_norm`, permitindo achar a posição em minúsculas/sem acento e recortar o
trecho no texto ORIGINAL, com acentuação e caixa preservadas para exibição.
"""
from __future__ import annotations

import re
import unicodedata
from dataclasses import dataclass

CONTEXTO_PADRAO = 160  # caracteres de cada lado do trecho casado

# Conectores intercambiáveis na redação jurídica ("cédula DO produto" vs "DE produto")
CONECTORES = frozenset({"de", "do", "da", "dos", "das", "no", "na", "nos", "nas",
                        "em", "a", "o", "as", "os", "e", "ou"})

# Marcas de vedação: o texto MENCIONA o ativo, mas para proibi-lo. Sinalizamos o
# trecho para o analista julgar — buscar menção não é o mesmo que buscar permissão.
NEGACOES = (
    "vedad", "vedar", "veda-se", "proibid", "proibir", "nao podera", "nao poderao",
    "nao sera permitid", "nao serao permitid", "nao autorizad", "impedid",
    "abster-se", "excluid", "nao podem", "nao pode",
)
JANELA_NEGACAO = 120  # caracteres antes do match onde a vedação é procurada


def fold_char(ch: str) -> str:
    """Um caractere -> um caractere: minúsculo e sem acento."""
    low = ch.lower()
    base = low[0] if low else ch  # 'İ'.lower() tem 2 chars; fica só o primeiro
    decomposed = unicodedata.normalize("NFD", base)
    return decomposed[0] if decomposed else base


def fold(text: str) -> str:
    return "".join(fold_char(c) for c in text)


def normalize_nfc(text: str) -> str:
    """Texto extraído é gravado em NFC para o fold preservar comprimento."""
    return unicodedata.normalize("NFC", text)


def tokens(termo: str) -> list[str]:
    return re.findall(r"[0-9a-z]+", fold(termo))


def _token_regex(token: str, flexivel: bool) -> str:
    if not flexivel:
        return re.escape(token)
    if token in CONECTORES:
        # "cédula DO produto" também casa "cédula DE produto"
        return f"(?:{'|'.join(sorted(CONECTORES))})"
    # plural nos dois sentidos: 'cédulas' no texto casa 'cédula' no termo e vice-versa
    raiz = token[:-1] if token.endswith("s") and len(token) > 3 else token
    return re.escape(raiz) + "s?"


def termo_regex(termo: str, flexivel: bool = True) -> re.Pattern[str]:
    """Regex tolerante a separadores: casa 'CPR-F', 'CPR F', 'CPR-\\nF', 'CPRF'.

    Os tokens precisam aparecer na ordem, ancorados em fronteira alfanumérica —
    'CPR-F' não casa dentro de 'CPR-Fulano'.

    Com `flexivel` (padrão), absorve as variações típicas da redação jurídica:
    plural e troca de conectores. `flexivel=False` exige a forma literal.
    """
    partes = tokens(termo)
    if not partes:
        raise ValueError(f"termo de busca sem conteúdo pesquisável: {termo!r}")
    corpo = r"[^0-9a-z]*".join(_token_regex(p, flexivel) for p in partes)
    return re.compile(rf"(?<![0-9a-z]){corpo}(?![0-9a-z])")


def token_prefiltro(termo: str) -> str:
    """Token mais longo do termo, usado no LIKE que reduz o universo no SQL.

    Devolve a raiz sem o plural e ignora conectores: o LIKE precisa ser mais
    permissivo que o regex de confirmação, nunca mais restrito, senão descarta
    documentos que o regex aceitaria.
    """
    partes = tokens(termo)
    if not partes:
        raise ValueError(f"termo de busca sem conteúdo pesquisável: {termo!r}")
    candidatos = [p for p in partes if p not in CONECTORES] or partes
    maior = max(candidatos, key=len)
    return maior[:-1] if maior.endswith("s") and len(maior) > 3 else maior


@dataclass
class Trecho:
    """Um trecho de texto com o pedaço casado localizado dentro dele."""

    texto: str
    match: str
    inicio: int  # posição do match dentro de `texto`
    posicao_documento: int  # offset do match no documento inteiro
    # o termo aparece logo após uma vedação ("vedada a aquisição de CPR-F"):
    # menção não é permissão, e quem lê o trecho precisa saber disso
    possivel_vedacao: bool = False

    @property
    def fim(self) -> int:
        return self.inicio + len(self.match)

    def destacado(self, abre: str = "«", fecha: str = "»") -> str:
        return f"{self.texto[: self.inicio]}{abre}{self.match}{fecha}{self.texto[self.fim :]}"


def detecta_vedacao(texto_norm: str, inicio: int, janela: int = JANELA_NEGACAO) -> bool:
    """Procura marca de vedação logo antes do termo encontrado."""
    contexto = texto_norm[max(0, inicio - janela) : inicio]
    contexto = re.sub(r"\s+", " ", contexto)
    return any(marca in contexto for marca in NEGACOES)


def _collapse(text: str) -> str:
    return re.sub(r"\s+", " ", text).strip()


def encontrar_trechos(
    texto: str,
    texto_norm: str,
    termos: list[str],
    *,
    max_trechos: int = 3,
    contexto: int = CONTEXTO_PADRAO,
    flexivel: bool = True,
) -> list[Trecho]:
    """Localiza os termos e devolve trechos recortados do texto original.

    `texto_norm` deve ter sido gerado por `fold(texto)` (mesmo comprimento).
    """
    achados: list[Trecho] = []
    vistos: set[int] = set()
    for termo in termos:
        for match in termo_regex(termo, flexivel=flexivel).finditer(texto_norm):
            inicio, fim = match.span()
            if inicio in vistos:
                continue
            vistos.add(inicio)
            antes = _collapse(texto[max(0, inicio - contexto) : inicio])
            casado = _collapse(texto[inicio:fim])
            depois = _collapse(texto[fim : fim + contexto])
            prefixo = f"{antes} " if antes else ""
            sufixo = f" {depois}" if depois else ""
            achados.append(
                Trecho(
                    texto=f"{prefixo}{casado}{sufixo}",
                    match=casado,
                    inicio=len(prefixo),
                    posicao_documento=inicio,
                    possivel_vedacao=detecta_vedacao(texto_norm, inicio),
                )
            )
            if len(achados) >= max_trechos:
                return achados
    return achados


def contem_termo(texto_norm: str, termos: list[str]) -> bool:
    return any(termo_regex(t).search(texto_norm) for t in termos)

File: web/lib/textos.ts
/**
 * Port da normalização e busca de trechos de `ingestor/textos.py`.
 *
 * As duas implementações precisam concordar: o Python grava `texto_norm` no
 * banco e o TypeScript recorta os trechos para exibição. `tests/textos.test.ts`
 * cobre os mesmos casos da suíte Python.
 *
 * Para evitar qualquer divergência de offset entre as linguagens (Python conta
 * code points, JS conta code units UTF-16), o `texto_norm` do banco é usado
 * apenas no filtro SQL; aqui o fold é recalculado a partir do texto original.
 */

export const CONTEXTO_PADRAO = 160;

export const CONECTORES = new Set([
  "de", "do", "da", "dos", "das", "no", "na", "nos", "nas",
  "em", "a", "o", "as", "os", "e", "ou",
]);

export const NEGACOES = [
  "vedad", "vedar", "veda-se", "proibid", "proibir", "nao podera", "nao poderao",
  "nao sera permitid", "nao serao permitid", "nao autorizad", "impedid",
  "abster-se", "excluid", "nao podem", "nao pode",
];

export const JANELA_NEGACAO = 120;

/** Um caractere -> um caractere: minúsculo e sem acento (preserva comprimento). */
export function foldChar(ch: string): string {
  const low = ch.toLowerCase();
  const base = low.length > 0 ? low[0] : ch;
  const decomposed = base.normalize("NFD");
  return decomposed.length > 0 ? decomposed[0] : base;
}

export function fold(text: string): string {
  let out = "";
  for (let i = 0; i < text.length; i++) out += foldChar(text[i]);
  return out;
}

export function tokens(termo: string): string[] {
  return fold(termo).match(/[0-9a-z]+/g) ?? [];
}

function escapeRegex(text: string): string {
  return text.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function tokenPattern(token: string, flexivel: boolean): string {
  if (!flexivel) return escapeRegex(token);
  if (CONECTORES.has(token)) return `(?:${[...CONECTORES].sort().join("|")})`;
  const raiz = token.endsWith("s") && token.length > 3 ? token.slice(0, -1) : token;
  return `${escapeRegex(raiz)}s?`;
}

/** Regex tolerante a separadores, plural e conectores, ancorada em fronteira. */
export function termoRegex(termo: string, flexivel = true): RegExp {
  const partes = tokens(termo);
  if (partes.length === 0) throw new Error(`termo sem conteúdo pesquisável: ${termo}`);
  const corpo = partes.map((p) => tokenPattern(p, flexivel)).join("[^0-9a-z]*");
  return new RegExp(`(?<![0-9a-z])${corpo}(?![0-9a-z])`, "g");
}

/**
 * Padrão para o filtro no PostgreSQL (regex POSIX, sem lookarounds).
 *
 * Usa `\y` (fronteira de palavra do Postgres) no lugar dos lookarounds. É uma
 * aproximação deliberadamente mais permissiva: o SQL nunca pode descartar o que
 * `termoRegex` aceitaria, e a confirmação final acontece sempre em JS.
 */
export function termoRegexSql(termo: string, flexivel = true): string {
  const partes = tokens(termo);
  if (partes.length === 0) throw new Error(`termo sem conteúdo pesquisável: ${termo}`);
  const corpo = partes.map((p) => tokenPattern(p, flexivel)).join("[^0-9a-z]*");
  return `\\y${corpo}\\y`;
}

/** Token mais longo, sem plural e ignorando conectores (prefiltro LIKE). */
export function tokenPrefiltro(termo: string): string {
  const partes = tokens(termo);
  if (partes.length === 0) throw new Error(`termo sem conteúdo pesquisável: ${termo}`);
  const candidatos = partes.filter((p) => !CONECTORES.has(p));
  const pool = candidatos.length > 0 ? candidatos : partes;
  const maior = pool.reduce((a, b) => (b.length > a.length ? b : a));
  return maior.endsWith("s") && maior.length > 3 ? maior.slice(0, -1) : maior;
}

export interface Trecho {
  texto: string;
  match: string;
  inicio: number;
  fim: number;
  posicaoDocumento: number;
  possivelVedacao: boolean;
}

function collapse(text: string): string {
  return text.replace(/\s+/g, " ").trim();
}

export function detectaVedacao(textoNorm: string, inicio: number, janela = JANELA_NEGACAO): boolean {
  const contexto = collapse(textoNorm.slice(Math.max(0, inicio - janela), inicio));
  return NEGACOES.some((marca) => contexto.includes(marca));
}

export function encontrarTrechos(
  texto: string,
  termos: string[],
  opcoes: { maxTrechos?: number; contexto?: number; flexivel?: boolean; textoNorm?: string } = {},
): Trecho[] {
  const { maxTrechos = 3, contexto = CONTEXTO_PADRAO, flexivel = true } = opcoes;
  const textoNorm = opcoes.textoNorm ?? fold(texto);
  const achados: Trecho[] = [];
  const vistos = new Set<number>();

  for (const termo of termos) {
    const rx = termoRegex(termo, flexivel);
    rx.lastIndex = 0;
    let match: RegExpExecArray | null;
    while ((match = rx.exec(textoNorm)) !== null) {
      const inicio = match.index;
      const fim = inicio + match[0].length;
      if (match[0].length === 0) {
        rx.lastIndex++;
        continue;
      }
      if (!vistos.has(inicio)) {
        vistos.add(inicio);
        const antes = collapse(texto.slice(Math.max(0, inicio - contexto), inicio));
        const casado = collapse(texto.slice(inicio, fim));
        const depois = collapse(texto.slice(fim, fim + contexto));
        const prefixo = antes ? `${antes} ` : "";
        const sufixo = depois ? ` ${depois}` : "";
        achados.push({
          texto: `${prefixo}${casado}${sufixo}`,
          match: casado,
          inicio: prefixo.length,
          fim: prefixo.length + casado.length,
          posicaoDocumento: inicio,
          possivelVedacao: detectaVedacao(textoNorm, inicio),
        });
        if (achados.length >= maxTrechos) return achados;
      }
    }
  }
  return achados;
}
